tests/test_cache.py

```diff
@@ -72,3 +73,29 @@ def test_get_cache_dir(monkeypatch):
     monkeypatch.setenv("TLDEXTRACT_CACHE", "/alt-tld-cache")
 
     assert get_cache_dir() == "/alt-tld-cache"
+
+
+def test_run_and_cache(tmpdir):
+    cache = DiskCache(tmpdir)
+
+    return_value1 = "unique return value"
+    some_fn = Mock(return_value=return_value1)
+    kwargs1 = {"value": 1}
+
+    assert some_fn.call_count == 0
+
+    call1 = cache.run_and_cache(some_fn, "test_namespace", kwargs1, kwargs1.keys())
+    assert call1 == return_value1
+    assert some_fn.call_count == 1
+
+    call2 = cache.run_and_cache(some_fn, "test_namespace", kwargs1, kwargs1.keys())
+    assert call2 == return_value1
+    assert some_fn.call_count == 1
+
+    kwargs2 = {"value": 2}
+    return_value2 = "another return value"
+    some_fn.return_value = return_value2
+
+    call3 = cache.run_and_cache(some_fn, "test_namespace", kwargs2, kwargs2.keys())
+    assert call3 == return_value2
+    assert some_fn.call_count == 2
```

Comment:
Thanks for the thorough PR! 🙌 

My only nitpick is, I had trouble grokking the test case. We're asserting side effects, so it's going to be an indirect test case no matter what. I expanded coverage of `run_and_cache`'s expected behavior with a `Mock` and asserting `call_count`.  I'm not sure which version I prefer. What do you think?

Reply:
I agree, I like your version more. 👍 
The only trade-off is that we're now testing for implementation details instead of side-effects, but it does make the test easier to understand.

Reply:
(Then again, by passing a function that can fail and asserting that it doesn't throw an exception I guess I was doing both, testing the implementation _and_ side-effects)